File: crypt1/crypt1.cpp
/*
  Author:wind1740
  LANG:C++
  PROG:crypt1
*/
#include <math.h>
#include <iostream>
#include <fstream>
using namespace std;

int n;
int map[100],mapp[100];
int sum;

	
bool checkk(int x)
{
  for (int i=1;i<=n;i++) 
  {
    if (map[i]==x)
		return (true);
  }
  return (false);
}

bool check(int x,int y)
{
  if (x>=pow (10,y))
	  return (false);
  if (x<pow (10,y-1))
	  return (false);
  int i=x;
  while (i!=0) 
  {
    if (checkk(i%10)==false)
		return (false);
    i=(int)i/10;
  }
  return (true);
}

void pro(int now)

{
  if (now==6) 
  {
    if (check((10*mapp[4]+mapp[5])*(100*mapp[1]+10*mapp[2]+mapp[3]),4)==true)
	{
      sum++;
	  
	}
  }
  else
  {
    for (int i=1;i<=n;i++) 
    {
      mapp[now]=map[i];
  	  if ((now==4) || (now==5))
	  {
	    if (check(mapp[now]*(100*mapp[1]+10*mapp[2]+mapp[3]),3)==true)
			pro(now+1);
      }
      else
	  {
        pro(now+1);
	  }
    }
  }
}
	
int main()
{
  ifstream in ("crypt1.in");
  ofstream out ("crypt1.out");
  in >> n;
  while (in.get()!='\n');
  for (int i=1;i<=n;i++)
    in >> map[i];  
  while (in.get()!='\n');
  sum=0;
  pro(1);
  
  out << sum <<endl;
  in.close ();
  out.close ();
}

File: crypt1/crypt1.pas
{
  Author:wind1740
  LANG:PASCAL
  PROG:crypt1
}
uses math;
var n:longint;
    map,mapp:array [1..100] of longint;
	sum:longint;
	
function checkk(x:longint):boolean;
var i:longint;
begin
  for i:=1 to n do 
  begin
    if map[i]=x then exit (true);
  end;
  exit (false);
end;

function check(x,y:longint):boolean;
var i:longint;
begin
  if x>=intpower (10,y) then exit (false);
  if x<intpower (10,y-1) then exit (false);
  i:=x;
  while i<>0 do
  begin
    if checkk(i mod 10)=false then exit (false);
    i:=i div 10;
  end;
  exit (true);
end;

procedure pro(now:longint);
var i:longint;
begin
  if now=6 then 
  begin
    if check((10*mapp[4]+mapp[5])*(100*mapp[1]+10*mapp[2]+mapp[3]),4)=true then
	begin
      inc (sum);
	  
	end;
  end
  else
  begin
    for i:=1 to n do 
    begin
      mapp[now]:=map[i];
  	  if (now=4) or (now=5) then
	  begin
	    if check(mapp[now]*(100*mapp[1]+10*mapp[2]+mapp[3]),3)=true then pro(now+1);
      end
      else
        pro(now+1);
    end;
  end;
end;
	
procedure init;
var i:longint;
begin
  readln (n);
  for i:=1 to n do
    read (map[i]);  
  sum:=0;
  pro(1);

  writeln (sum);
end;

begin
  assign (input,'crypt1.in');
  assign (output,'crypt1.out');
  reset (input);
  rewrite (output);
  
  init;
  
  close (input);
  close (output);
end.